### DISCUSSION TITLE
Problems with compilation failures in version 3.7.9

### DISCUSSION THREAD

**Discussion by YamatoMedana** [zeroc-ice/ice#1500] [General]
### Version

Ice version:3.7.9
OS:linux(ubuntu 22.04) 

### Describe the bug

When I compile the source code for the cpp section, I get the error shown in the figure. I think I have installed the relevant third-party library and declared the path in the environment variable. However, when I do "make", there are some errors, I guess should be source code problems? Can someone help me with this?
![3)WZ2 KNW))GR55$NC`S430](https://github.com/zeroc-ice/ice/assets/143420380/9a3ef223-ad21-4095-a05e-2f043f9eaa53)

### Expected behavior

Hope the compilation can be completed smoothly

**Comment by externl** [comment]
Hi @YamatoMendana, a few questions

- Which environment variables are you setting? 
- Which version of mcpp version are you using? Ice relies on a patched version of mcpp.

**Comment by YamatoMedana** [comment]
I have set the basic third-party software in /usr/local/... ,
As shown in the figure. And I also set the corresponding path under the config folder make.rules, but still can not.

My mcpp version is as follows
MCPP V.2.7.2 (2008/11) compiler-independent-build compiled by GCC V.11.4

Of course, I also tried to install mcpp according to the official apt and compile the ice source code, but I still couldn't.

![LK7BJ760P6CM1WK(33W6 1](https://github.com/zeroc-ice/ice/assets/143420380/32ef00f8-dca2-4025-89db-87bd3cc7146c)

**Comment by pepone** [comment]
Hi,

To build Ice you need the mcp patched version, Ubuntu package includes the required patches specially [01-zeroc-fixes.patch](https://salsa.debian.org/debian/mcpp/-/blob/master/debian/patches/01-zeroc-fixes.patch).

What errors do you get when building with MCPP from the Ubuntu repository, I assume it will be a different issue because the error you posted is related to an issue fixed in the linked patch.

**Comment by YamatoMedana** [comment]
> Hi,
> 
> To build Ice you need the mcp patched version, Ubuntu package includes the required patches specially [01-zeroc-fixes.patch](https://salsa.debian.org/debian/mcpp/-/blob/master/debian/patches/01-zeroc-fixes.patch).
> 
> What errors do you get when building with MCPP from the Ubuntu repository, I assume it will be a different issue because the error you posted is related to an issue fixed in the linked patch.

emmmm..... Your patch I used, but there is no effect (even did not generate the corresponding library), I think it is still the source code itself. Can not identify 'Identity', even said 'IceStorm' the same name. Although I do not know how to modify, but intuition tells me: this is a source code problem.......

**Comment by pepone** [comment]
Hi,

We build Ice regularly in Ubuntu and Debian using the mcpp packages provided by the distribution, and it should work out of the box.

It is simpler to use the package provided by ZeroC  see https://zeroc.com/downloads/ice/3.7/cpp#ubuntu

The same sources are also used to build the official Ubuntu packages from https://packages.ubuntu.com/source/jammy/zeroc-ice.

**Comment by YamatoMedana** [comment]
> Hi,
> 
> We build Ice regularly in Ubuntu and Debian using the mcpp packages provided by the distribution, and it should work out of the box.
> 
> It is simpler to use the package provided by ZeroC see https://zeroc.com/downloads/ice/3.7/cpp#ubuntu
> 
> The same sources are also used to build the official Ubuntu packages from https://packages.ubuntu.com/source/jammy/zeroc-ice.
I have tried, using the installation method provided by the website, it is feasible, but the reason I compile the source code is because I need the corresponding library and header files. So, I haven't been able to solve the problem of source code compilation errors, maybe I should try to compile the lower version

**Comment by pepone** [comment]
> I compile the source code is because I need the corresponding library and header files. 

The library and header files are included in the `libzeroc-ice-dev` package, see the package contents in https://packages.ubuntu.com/jammy/amd64/libzeroc-ice-dev/filelist

**Comment by YamatoMedana** [comment]
> > I compile the source code is because I need the corresponding library and header files.
> 
> The library and header files are included in the `libzeroc-ice-dev` package, see the package contents in https://packages.ubuntu.com/jammy/amd64/libzeroc-ice-dev/filelist

Thanks for your help, I found the location of the corresponding library and started to get familiar with the api usage smoothly. However, I am still very tangled, why I can not compile the source code in ubuntu22, is there really not acclimation? Hopefully there will be a patch that will save me

**Comment by pepone** [comment]
>  I am still very tangled, why I can not compile the source code in ubuntu22, is there really not acclimation? Hopefully there will be a patch that will save me

You certainly can, but need to use the correct 3rd party packages, regarding MCPP there are several possibilities:

- Use the mcpp package from Ubuntu repositories, in this case, you should not set MCPP_HOME environment variables
- Or, build mcpp from zeroc repository https://github.com/zeroc-ice/mcpp, which includes the required patches, you must set MCPP_HOME to the corresponding MCPP installation directory when building Ice
- Or, build mcpp from the upstream repository, after applying ZeroC patches, you must set MCPP_HOME to the corresponding MCPP installation directory when building Ice

**Comment by YamatoMedana** [comment]
The current environment is ubuntu22.04 (i386)



It's finally done.

Among them, it is not feasible to download the relevant third-party library compression package (maybe I am not correct about "make" and related configuration), but download the relevant third-party compression package (such as mcpp? bzip2?) Unzip, configure 'confiure' and 'make', and then make the cpp part of ice3.7.9.



Finally, the correct installation method:

Step 1: Install ice third-party library dependencies



apt-get install libmcpp-dev



apt-get install libssl-dev



apt-get install liblmdb-dev



apt-get install libexpat1-dev



apt-get install libbz2-dev



apt-get install dpkg-dev



apt-get install libedit-dev



Part Two: Configuration



cd ice-3.7.9/config



vim make.rules



Find prefix=?? (Set where you need to install it)

![image](https://github.com/zeroc-ice/ice/assets/143420380/79a2e6a4-0865-4860-be36-3efab90c8855)


Part 3: Compiling ice



cd ice-3.7.9/cpp



make



make install

![image](https://github.com/zeroc-ice/ice/assets/143420380/ac01399f-0487-423b-9425-12652b8553fa)


Step 4: Set the environment variables



Step 5: Verify the installation is successful

slice2pp -v

 
![image](https://github.com/zeroc-ice/ice/assets/143420380/3086255d-1c78-41f8-8f3f-1d376c4b847c)
